packages/spacecat-shared-rum-api-client/src/functions/rageclick.js

```diff
@@ -0,0 +1,179 @@
+    if (checkpoint === 'click' && !selectorsIgnoreList.includes(source)) {
+      if (!clickSelectors[source]) {
+        clickSelectors[source] = 0;
+      }
+      clickSelectors[source] += 1;
+    }
+  }
+  for (const selector of Object.keys(clickSelectors)) {
+    if (clickSelectors[selector] < threshold) {
+      delete clickSelectors[selector];
+    }
+  }
+  return clickSelectors;
+}
+
+function filterRageClickInstancesByThreshold(
+  rageClickInstances,
+  pageData,
+  rageClickPercentThreshold,
+  rageClickPageviewThreshold,
+) {
+  for (const url of Object.keys(rageClickInstances)) {
+    if (pageData[url].pageViews < rageClickPageviewThreshold) {
+      // eslint-disable-next-line no-param-reassign
+      delete rageClickInstances[url];
+    } else {
+      for (const selector of Object.keys(rageClickInstances[url])) {
+        const rageClickPercentage = (
+          rageClickInstances[url][selector].samples / pageData[url].samples) * 100;
+        if (rageClickPercentage < rageClickPercentThreshold) {
+          // eslint-disable-next-line no-param-reassign
+          delete rageClickInstances[url][selector];
+        } else {
+          // eslint-disable-next-line no-param-reassign
+          rageClickInstances[url][selector].percentage = rageClickPercentage;
+        }
+      }
+      if (Object.keys(rageClickInstances[url]).length === 0) {
+        // eslint-disable-next-line no-param-reassign
+        delete rageClickInstances[url];
+      } else {
+        // eslint-disable-next-line no-param-reassign
+        rageClickInstances[url].pageViews = pageData[url].pageViews;
+        // eslint-disable-next-line no-param-reassign
+        rageClickInstances[url].samples = pageData[url].samples;
+      }
+    }
+  }
+}
+
+function getRageClickOpporunities(rageClickInstances) {
+  const opportunities = [];
+  for (const url of Object.keys(rageClickInstances)) {
+    const opportunity = {
+      type: OPPORTUNITY_TYPE,
+      page: url,
+      screenshot: '',
+      trackedPageKPIName: OPPORTUNITY_DESCRIPTION,
+      trackedPageKPIValue: '',
+      pageViews: rageClickInstances[url].pageViews,
+      samples: rageClickInstances[url].samples,
+      metrics: [],
+    };
+    for (const selector of Object.keys(rageClickInstances[url])) {
+      if (typeof rageClickInstances[url][selector] === 'object') {
+        opportunity.metrics.push({
+          type: 'click',
+          selector,
+          value: rageClickInstances[url][selector].value,
+          samples: rageClickInstances[url][selector].samples,
+          percentage: rageClickInstances[url][selector].percentage,
+        });
+      }
+    }
+    const avgRageClickPercentage = opportunity.metrics.reduce(
+      (acc, metric) => acc + metric.percentage,
+      0,
+    ) / opportunity.metrics.length;
+    opportunity.trackedPageKPIValue = avgRageClickPercentage;
+    opportunities.push(opportunity);
+  }
+  return opportunities;
+}
+
+function handler(bundles) {
```

Comment:
_personal opinion and not blocking_: I don't know how much memory overhead it would add but using functional js interfaces `map`, `filter` etc. would increase the readability

Reply:
as I was building a global structure `rageClickInstances`, felt like map wasn't best fit and running multiple operations of map/filter, will iterate the array more than once, so I thought its better to create the structure I want with single iteration.